### PR TITLE
[PD] fix pad uptoface and uptoshape

### LINKED ISSUE #14663: PartDesign: Pad up to face fails in some scenarios
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

This is a tricky one and hard to replicate, so I have attached a model I created where the problem exists. This model is fine in 0.21.

The problem can be brought about by loading the model, then  changing the Binder001 in Body Cover to frozen, or detached, essentially making a change to the model. At this point Sketch 017 will  fail. There are various issues, but the one that seems most obvious is that the Pad MountPosts fails. This pad uses up to face as the type and rather than padding up to the nominated face, it appears to pad FROM the nominated face to infinity. I tried to replicate this scenario with a simpler two body model, but naturually it worked, so I included my file so whomever looks at it can see exactly what is happening to cause the issue.

I initially drew this in Onsdel, but then I updated Onsdel to the latest dev build and it has the same issue as Freecad 0.22. In Freecad 0.21 it's all good. 

### Full version info

```shell
OS: Windows 11 build 22631
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.37709 (Git)
Build type: Release
Branch: main
Hash: dd816a635ccacc39f29b1761c6ccfed21a90b646
Python 3.11.9, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: English/Australia (en_AU)
Installed mods: 
  * Curves.backup1706353084.2536438 0.6.13
  * fasteners.backup1718262130.4318058 0.5.22
  * OpenDark.backup1718262281.5371034 2024.3.13
```

### Subproject(s) affected?

PartDesign

### Anything else?

[BoseQC15Headset.zip](https://github.com/user-attachments/files/15831328/BoseQC15Headset.zip)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #16557: Part Design, Pocket, Type: Up to face, with Offset Bug
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

There is a bug when offsetting a Part Design, Pocket feature with Type: "Up to face" and the face selected is the inner face of half of 
section of a tube. The offset is applied to the outer face of the tube section, instead of the selected, inner face. 

Steps to reproduce:
Using the Part Design workflow:
1. Create a body that is half of a section of a tube (see attached example for reference).
2. Create a sketch of the shape to be debossed on the inner face of the tube section.
3. Create a Pocket feature, setting "Type" to "Up to face", selecting the inner face.
4. Set "Offset to face" to some value less than the wall thickness of the pipe section.
In my testing, the offset will appear on the opposite face.

Example file:
[Pocket Up to Face Bug.zip](https://github.com/user-attachments/files/17006537/Pocket.Up.to.Face.Bug.zip)

### Full version info

```shell
OS: Windows 11 build 22631
Word size of FreeCAD: 64-bit
Version: 1.0.0RC1.38684 (Git)
Build type: Release
Branch: main
Hash: 92e31fa7ea974fe48c372f6749916d68f42dcf95
Python 3.11.9, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: OpenLight.qss/OpenLight/
Installed mods: 
  * Curves 0.6.47
  * Ondsel-Lens 2024.7.5.02
  * OpenTheme 2024.9.1
```

### Subproject(s) affected?

PartDesign

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by FlachyJoe**
Fix #14663 
Fix #16557

### CODE REVIEW

**Comment by chennes** [comment]
Anything I can help with here, @FlachyJoe ? I'm working on killing blockers today.

**Comment by FlachyJoe** [comment]
@chennes It works with the base but has to be checked for another shape e.g. a Binder.

**Comment by FlachyJoe** [comment]
The fix works for the base shape but not for some other, it raises an OCC error.
To treat the other shapes I've now the choice to:

- keep the current implementation which doesn't produce the attempted result in some cases
- produce an OCC error by applying the same process than to the base. Useless work but informs the user easily.
- try to detect the unattempted result (which pass the validity check) and inform the user

I think we can merge this PR now because it improves the feature anyway. The other failing cases can be fixed in another commit.

**Comment by FEA-eng** [comment]
I compiled it and checked with your simplified file (the file attached to the issue description has errors with sketches after recomputing). Up to face still goes to outer space but up to shape works even when I add a second circle somewhere above the original one.

**Comment by FlachyJoe** [comment]
@FEA-eng You're right, I lost something during update. 🤔

**Comment by FlachyJoe** [comment]
@bgbsww  @realthunder I find this code in `TopoShape::makeElementPrismUntil`

```cpp
    TopoShape _uptoface(__uptoface);
    if (checkLimits && _uptoface.shapeType(true) == TopAbs_FACE
        && !BRep_Tool::NaturalRestriction(TopoDS::Face(_uptoface.getShape()))) {
        // When using the face with BRepFeat_MakePrism::Perform(const TopoDS_Shape& Until)
        // then the algorithm expects that the 'NaturalRestriction' flag is set in order
        // to work as expected.
        BRep_Builder builder;
        _uptoface = _uptoface.makeElementCopy();
        builder.NaturalRestriction(TopoDS::Face(_uptoface.getShape()), Standard_True);
    }
```

Do you know where come this info from? I can't find any mention of `NaturalRestriction` in `BRepFeat_MakePrism` code and if I comment these lines then UpToFace works better. :exclamation:  :question:

**Comment by bgbsww** [comment]
I do not know where that came from, and reading the OCCT source code does not directly support that statement.

It is the case that that flag does exist, and the two explaining remarks as as follows:

  //! Returns Standard_True if the face is not trimmed.

and

  //! Returns TRUE if the boundary of this face is known to be the parametric space (Umin, UMax, VMin, VMax).

To my reading that would mean the setting the builder.NaturalRestriction True on a face that isn't larger than
 the intersection of the padded object would tend to cause bugs.

Is this failing in cases where the UpToFace is either too small, or exactly the size with a tolerance error?

**Comment by realthunder** [comment]
This comment was added by @wwmayer if I remember correctly. And if you remove this, many simple case will break. Removing might work for some cases, especially if concave up to face is involved, but that's just a manifestation of the OCCT buggy operation. There will always be failure cases unless we fix it in OCCT. You can search my branch for function of the same name to see my not so successful attempts.

**Comment by FlachyJoe** [comment]
> There will always be failure cases unless we fix it in OCCT.

I agree but OCC fixing workflow seems very slow and we have to workaround for next years.

If this trick doesn't work in some cases we should disable it in these.

**Comment by wwmayer** [comment]
For your record:
This commit mentions the flag: bc5311b3eec28b2c
This is the related bug report: https://tracker.freecad.org/view.php?id=4567
This is the forum thread: https://forum.freecad.org/viewtopic.php?f=8&t=55603

File: src/Mod/Part/App/TopoShapeExpansion.cpp [hunk 1]
```diff
@@ -4237,9 +4237,33 @@ TopoShape& TopoShape::makeElementPrismUntil(const TopoShape& _base,
...
+            isConcave = dirdU.Angle(direction) < M_PI_2;
+            gp_Dir dirdV(dV);
+            isConcave = isConcave || (dirdV.Angle(direction) < M_PI_2);
+        }
+
+        if (! isConcave){
+            BRep_Builder builder;
+            _uptoface = _uptoface.makeElementCopy();
+            builder.NaturalRestriction(TopoDS::Face(_uptoface.getShape()), Standard_True);
+        }
```
Comment (FlachyJoe) [inline src/Mod/Part/App/TopoShapeExpansion.cpp:4266]: @realthunder what do you think about this? It works for [my test case](https://github.com/FreeCAD/FreeCAD/issues/14663#issuecomment-2166651538) but you could imagine a failure scenario.

**Comment by FlachyJoe** [comment]
@chennes could you help with the windows build failure? I haven't access to a windows system to test.

**Comment by chennes** [comment]
> @chennes could you help with the windows build failure? I haven't access to a windows system to test.

Yeah, it's failing because `windows.h` defines `near` and `far` as macros (once again demonstrating the evils of macros!). I just renamed your variables, it works fine now.

**Comment by bgbsww** [comment]
Is there any chance of a script of the test case, that could maybe be altered to show one of @realthunder failure cases working?

**Comment by chennes** [comment]
@AjinkyaDahale is going to record a macro of the failure for test generation.

**Comment by FlachyJoe** [comment]
I've completed the fix for #16557 but I broke something :hammer: :cry: 

Is it repairable or does it need a new PR?

**Comment by AjinkyaDahale** [comment]
> I've completed the fix for #16557 but I broke something with the git branch 🔨 😢
> 
> Is it repairable or does it need a new PR?

A git reply may fix it.

**Comment by AjinkyaDahale** [comment]
Macro added in issue.

**Comment by chennes** [comment]
Cherry-picked to 1.0

**Pull request merged**